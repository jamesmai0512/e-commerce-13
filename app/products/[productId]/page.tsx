import dynamic from 'next/dynamic'

// Component
const Button = dynamic(() => import('@/components/Common/Button'))
const ProductDetail = dynamic(
  () => import('@/components/Product/ProductDetail'),
)

const Input = dynamic(() => import('@/components/Common/Input'))
const BoxInfo = dynamic(() => import('@/components/Common/Box/BoxInfo'))

// Constants
import { API, ORDERSTEPS, TProduct } from '@/constants/common'

import { getProduct } from '@/utils/fetchers/ProductFetcher'
import styles from '../Product.module.css'

export interface ProductDetailPageProps {
  params: { productId: string }
}

async function generateStaticParams() {
  const products = await fetch(`${API}/products`).then((res) => res.json())

  return products.map((product: TProduct) => ({
    productId: product.id,
  }))
}

const ProductDetailPage = async ({
  params,
}: {
  params: { productId: string }
}) => {
  const {
    introduce_brand,
    order_step,
    email_form,
    form_text,
    form_text_title,
    form_text_description,
    form_email_submit,
  } = styles

  const { productId } = params
  const product = await getProduct(productId)

  return (
    <div>
      <ProductDetail product={product} />
      <h1 className={introduce_brand}>What makes our brand different</h1>
      <div className={order_step}>
        {ORDERSTEPS.map((i) => (
          <BoxInfo
            key={i.id}
            icon={i.icon}
            title={i.title}
            description={i.description}
          />
        ))}
      </div>

      <div className={email_form}>
        <div className={form_text}>
          <h1 className={form_text_title}>
            Join the club and get the benefits
          </h1>
          <p className={form_text_description}>
            Sign up for our newsletter and receive exclusive offers on new
            ranges, sales, pop up stores and more
          </p>
          <form className={form_email_submit}>
            <Input background="light" placeholder="your@email.com" />
            <Button background="purple" text="Sign Up" />
          </form>
        </div>
      </div>
    </div>
  )
}

export default ProductDetailPage
