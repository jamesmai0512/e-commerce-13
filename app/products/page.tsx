import ProductList from '@/components/Product/ProductList'
import LoadingIndicator from '@/components/Common/LoadingIndicator'
import { getProductList } from '@/utils/fetchers/ProductFetcher'

import styles from './Product.module.css'

const ProductsPage = async () => {
  const { none_product, text_alert_info, text_alert } = styles
  const products = await getProductList()

  console.log(products)

  if (!products) return <LoadingIndicator />

  return (
    <div>
      <div>
        {products && products.length ? (
          <ProductList products={products} />
        ) : (
          <div className={none_product}>
            <div className={text_alert_info}>
              <h1 className={text_alert}>There are no products!!</h1>
            </div>
          </div>
        )}
      </div>
    </div>
  )
}

export default ProductsPage
