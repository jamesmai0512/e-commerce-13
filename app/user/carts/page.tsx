// Component
import CartList from '@/components/Cart/CartList'

// Fetcher
import { getCarts } from '@/utils/fetchers/CartFetcher'

import styles from './Carts.module.css'

const { none_product, text_alert_info, text_alert } = styles

export interface CartItem {
  id: string
  quantity: number
  productId: number
  price: number
}

const Carts = async () => {
  const carts = await getCarts()

  return (
    <div>
      {carts && carts.length ? (
        <CartList carts={carts} />
      ) : (
        <div className={none_product}>
          <div className={text_alert_info}>
            <h1 className={text_alert}>There are no carts!!</h1>
          </div>
        </div>
      )}
    </div>
  )
}

export default Carts
