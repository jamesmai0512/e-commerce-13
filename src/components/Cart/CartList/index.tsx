'use client'

import { Suspense, useCallback, useMemo, useState } from 'react'
import dynamic from 'next/dynamic'

// Components
const Button = dynamic(() => import('@/components/Common/Button'))
const Cart = dynamic(() => import('@/components/Cart'))

import styles from './CartList.module.css'

// type
import { CartItem } from '@/app/user/carts/page'
import Loading from '@/app/loading'
interface ICartListProps {
  carts: CartItem[]
}

const CartList = ({ carts }: ICartListProps) => {
  const [cartItems, setCartItems] = useState<CartItem[]>(carts)

  const {
    shopping_cart,
    shopping_cart_title,
    shopping_cart_info,
    shopping_cart_info_product,
    shopping_cart_list,
    shopping_cart_total,
    shopping_cart_total_content,
    shopping_cart_list_detail,
    shopping_cart_list_subtotal,
    shopping_cart_list_price,
    shopping_carts_list,
  } = styles

  const updateCartItemQuantity = useCallback(
    (cartItemId: string, newQuantity: number) => {
      if (cartItems) {
        const updatedCartItems = cartItems.map((cartItem: CartItem) =>
          cartItem.id === cartItemId
            ? { ...cartItem, quantity: newQuantity }
            : cartItem,
        )
        setCartItems(updatedCartItems)
      }
    },
    [cartItems],
  )

  const deleteCartItem = useCallback(
    (cartItemId: string) => {
      if (cartItems) {
        const updatedCartItems = cartItems.filter(
          (cartItem) => cartItem.id !== cartItemId,
        )
        setCartItems(updatedCartItems)
      }
    },
    [cartItems],
  )

  const totalPrice = useMemo(() => {
    if (cartItems) {
      return cartItems.reduce(
        (acc, cartItem) => acc + cartItem.quantity * cartItem.price,
        0,
      )
    }
  }, [cartItems])

  return (
    <Suspense fallback={<Loading />}>
      <div className={shopping_cart}>
        <h1 className={shopping_cart_title}>Your shopping cart</h1>

        <div className={shopping_cart_info}>
          <p className={shopping_cart_info_product}>Product</p>
          <p className="table_name">Quantity</p>
          <p className="table_name">Total</p>
        </div>
        <div className={shopping_cart_list}>
          {cartItems.map((cartItem) => (
            <Cart
              key={cartItem.id}
              productId={cartItem.productId}
              cartId={cartItem.id}
              quantity={cartItem.quantity}
              price={cartItem.price}
              updateCartItemQuantity={updateCartItemQuantity}
              deleteCartItem={deleteCartItem}
            />
          ))}
        </div>
        <div className={shopping_cart_total}>
          <div className={shopping_cart_total_content}>
            <div className={shopping_cart_list_detail}>
              <h3 className={shopping_cart_list_subtotal}>Subtotal</h3>
              <h2 className={shopping_cart_list_price}>Â£{totalPrice}</h2>
            </div>
            <p className={shopping_carts_list}>
              Taxes and shipping are calculated at checkout
            </p>
            <Button background="purple" text="Go to checkout" />
          </div>
        </div>
      </div>
    </Suspense>
  )
}

export default CartList
