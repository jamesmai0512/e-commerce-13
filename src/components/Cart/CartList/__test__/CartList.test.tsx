import { CartItem } from '@/app/user/carts/page'
import { render, fireEvent } from '@testing-library/react'
import CartList from '../'

describe('CartList', () => {
  const cartItems: CartItem[] = [
    { id: '1', productId: 1, quantity: 1, price: 10 },
    { id: '2', productId: 2, quantity: 2, price: 20 },
  ]

  it('Should render correctly', () => {
    const component = render(<CartList carts={cartItems} />)
    expect(component).toMatchSnapshot()
  })
})
