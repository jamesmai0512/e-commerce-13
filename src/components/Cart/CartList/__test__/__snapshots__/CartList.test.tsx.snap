// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartList Should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="shopping_cart"
      >
        <h1
          class="shopping_cart_title"
        >
          Your shopping cart
        </h1>
        <div
          class="shopping_cart_info"
        >
          <p
            class="shopping_cart_info_product"
          >
            Product
          </p>
          <p
            class="table_name"
          >
            Quantity
          </p>
          <p
            class="table_name"
          >
            Total
          </p>
        </div>
        <div
          class="shopping_cart_list"
        />
        <div
          class="shopping_cart_total"
        >
          <div
            class="shopping_cart_total_content"
          >
            <div
              class="shopping_cart_list_detail"
            >
              <h3
                class="shopping_cart_list_subtotal"
              >
                Subtotal
              </h3>
              <h2
                class="shopping_cart_list_price"
              >
                £
                50
              </h2>
            </div>
            <p
              class="shopping_carts_list"
            >
              Taxes and shipping are calculated at checkout
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="shopping_cart"
    >
      <h1
        class="shopping_cart_title"
      >
        Your shopping cart
      </h1>
      <div
        class="shopping_cart_info"
      >
        <p
          class="shopping_cart_info_product"
        >
          Product
        </p>
        <p
          class="table_name"
        >
          Quantity
        </p>
        <p
          class="table_name"
        >
          Total
        </p>
      </div>
      <div
        class="shopping_cart_list"
      />
      <div
        class="shopping_cart_total"
      >
        <div
          class="shopping_cart_total_content"
        >
          <div
            class="shopping_cart_list_detail"
          >
            <h3
              class="shopping_cart_list_subtotal"
            >
              Subtotal
            </h3>
            <h2
              class="shopping_cart_list_price"
            >
              £
              50
            </h2>
          </div>
          <p
            class="shopping_carts_list"
          >
            Taxes and shipping are calculated at checkout
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
